### DISCUSSION TITLE
spacy caching explanation needed

### DISCUSSION THREAD

**Discussion by saraswat40** [explosion/spaCy#10758] [Help: Coding & Implementations] [answered]
I have a very simple program:
```
import time
import spacy

start = time.time()
nlp = spacy.load('en_core_web_lg', disable=['tagger', 'ner', 'textcat', '...'])
print(time.time() - start)
```
The first time I run this I get this output:
```
# python3 test.py 
183.59882807731628

```
The second time I run this I get this:
```
# python3 test.py 
1.3441166877746582
```
Obviously some sort of caching is happening. Is there an explanation for this somewhere?

My real problem is that I have multiple threads accessing a file like this. For the first invocation, they all get stuck trying to load spacy. My solution is to create a lockfile and the lock acquirer will load spacy and rest of them will wait and use the cache mentioned above. Is there a better way to do this?

Thanks

**Comment by polm** [comment, marked as answer]
spaCy has no particular features related to caching between separate processes. I suspect what is happening here is that the related files are cached by your OS's disk handling layer, but I am not sure of that. In general I wouldn't expect the difference to be that huge though.

**Comment by saraswat40** [comment]
Hello, we also loading the universal sentence encoder using hub.load. That one doesn't show this behavior. Thats why I thought this is spacy specific.

**Reply by saraswat40** [reply]
Can you confirm that under the covers, spacy.load is just f.open and whatever else it needs to do to load the model and there is no internal caching mechanism implemented?

**Reply by polm** [reply]
spaCy has no persistence between processes and isn't doing anything clever.

If you are loading a Transformers model, it may have to be downloaded the first time. That doesn't apply to the large model in your example script. You should also see console output that indicates the download is happening if that's the case.

**Comment by saraswat40** [comment]
Thanks for clarifying. The transformer model is also loaded from local disk.
